lib/vsdata.cpp

```diff
@@ -47,6 +47,60 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int order = gf.FESpace()->GetMaxElementOrder();
+
+   // check for integral elements
+   const int dim = mesh->Dimension();
+   const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+   if (fec && fec->GetMapType(dim) == FiniteElement::INTEGRAL)
+   {
+      cout << "Warning: integral elements are non-polynomial in the physical space,\n"
+           << "         consider increasing the refinement by the key 'o'."
+           << endl;
+   }
+
+   return std::max(order, 1);
+}
+
+int VisualizationSceneScalarData::GetAutoRefineFactor()
+{
+   const int dim = mesh->Dimension();
+   const int ne = (dim == 3)?(mesh->GetNBE()):(mesh->GetNE());
+
+   // determine the refinement based on the order of the mesh and grid function
+   int order_ref = GetFunctionAutoRefineFactor();
+
+   // mesh
+   const FiniteElementSpace *nfes = mesh->GetNodalFESpace();
+   if (nfes)
+   {
+      const int order = nfes->GetMaxElementOrder();
+      order_ref = std::max(order_ref, order);
+   }
+
+   // limit the total number of elements
+   int auto_ref_surf_elem = ne*(order_ref+1)*(order_ref+1);
+   auto_ref_surf_elem = std::min(std::max(auto_ref_surf_elem,
+                                          auto_ref_min_surf_elem), auto_ref_max_surf_elem);
+
+   // approach the given number of elements
+   int ref = 1;
+   while (ref < auto_ref_max && ne*(ref+1)*(ref+1) <= auto_ref_surf_elem)
+   { ref++; }
+
+   if (ref < order_ref)
+   {
+      cout << "Warning: the automatic refinement does not resolve the data fully,\n"
```

Comment:
Might be helpful to mention some of the reasoning why. E.g. something such as

The automatic refinement tried to refine X times but could only refine Y times without exceeding the soft-threshold for maximum elements: `auto_ref_max_surf_elem = Z`. Consider overriding by manually increasing refinement using the key 'o'.

Reply:
It shows the automatic refinement level, so this is just complimentary warning to it. I would keep it short.

Reply:
Besides that order_ref does not guarantee anything, so giving it to the user is misleading potentially. The user should check visually if he/she is satisfied with the result. The warning just reminds him/her to do so, because what is on the screen might be very off.